pkg/api/grids/grids_test.go

```diff
@@ -0,0 +1,142 @@
+			assert.Equal(t, "test-coherence-image:latest", myGrids[0].Image, "expect Grid Image Name to be test-coherence-image:latest")
+		})
+	}
+}
+
+// TestReturnSingleGrid tests that the Grid with ID=0 is returned
+// GIVEN that Grid with ID=0 exists
+// WHEN I call ReturnSingleGrid
+// THEN all the Grid with ID=0 should be returned
+// AND that the Grid returned should have the correct ID and have valid values
+func TestReturnSingleGrid(t *testing.T) {
+	// GIVEN empty model binding pairs, managed clusters and fake kubernetes clients.
+	var clients kubernetes.Interface = k8sfake.NewSimpleClientset()
+	clusters := testutil.GetTestClusters()
+	type args struct {
+		w http.ResponseWriter
+		r *http.Request
+	}
+	tests := []struct {
+		name    string
+		listers controller.Listers
+	}{
+		{
+			name:    "verifyReturnSingleGrid",
+			listers: testutilcontroller.NewControllerListers(&clients, clusters, &modelBindingPairs),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			vars := map[string]string{
+				"id": "0",
+			}
+			var myGrid Grid
+			Init(tt.listers)
+			resp := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/grids", nil)
+			req = mux.SetURLVars(req, vars)
+			ReturnSingleGrid(resp, req)
+			assert.Equal(t, http.StatusOK, resp.Code, "expect the http return code to be http.StatusOk")
+			json.NewDecoder(resp.Body).Decode(&myGrid)
+			assert.Equal(t, "0", myGrid.ID, "expect the Grid with ID = 0 to be successfully returned")
+			assert.Equal(t, "test-coherence", myGrid.Name, "expect Grid Name to be test-coherence")
+			assert.Equal(t, "test-coherence-image:latest", myGrid.Image, "expect Grid Image Name to be test-coherence-image:latest")
+		})
+	}
+}
+
+// TestReturnSingleGridNotFound tests that a Grid that doesn't exist is not found
+// GIVEN that Grid with ID=100 doesn't exist
+// WHEN I call ReturnSingleGrid
+// THEN all no Grid should be returned
+// AND that the Grid returned should be empty
+func TestReturnSingleGridNotFound(t *testing.T) {
+	// GIVEN empty model binding pairs, managed clusters and fake kubernetes clients.
+	var clients kubernetes.Interface = k8sfake.NewSimpleClientset()
+	clusters := testutil.GetTestClusters()
+	type args struct {
+		w http.ResponseWriter
+		r *http.Request
+	}
+	tests := []struct {
+		name    string
+		listers controller.Listers
+	}{
+		{
+			name:    "verifyReturnSingleGridNotFound",
+			listers: testutilcontroller.NewControllerListers(&clients, clusters, &modelBindingPairs),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			vars := map[string]string{
+				"id": "100",
+			}
+			var myGrid Grid
+			Init(tt.listers)
+			resp := httptest.NewRecorder()
+			req := httptest.NewRequest("GET", "/grids", nil)
+			req = mux.SetURLVars(req, vars)
+			ReturnSingleGrid(resp, req)
+			assert.Equal(t, http.StatusOK, resp.Code, "expect the http return code to be http.StatusOk")
```

Comment:
the API should be fixed to make this return 404

Reply:
Is that part of it?
I did not know we are supposed to fixing the APIs cause they all do that

Reply:
In this case, it's so bad and wrong, that some of them have been fixing as we go. My view is that while you are there is a good time to fix it. Coming back to fix all of them later, no one is going to remember everything that needs fixing